File: Greedy/Python/Jump Game II.py
# Problem link: https://leetcode.com/problems/jump-game-ii/description/
# Time: O(n)
# Space: O(1)
class Solution:
    def jump(self, nums: List[int]) -> int:
        l = r = res = 0

        while r < len(nums) - 1:
            far = 0
            for i in range(l, r + 1):
                far = max(far, i + nums[i])
            l, r, res = r + 1, far, res + 1
        
        return res

File: Greedy/Java/Jump Game II.java
// Problem link: https://leetcode.com/problems/jump-game-ii/description/
// Time: O(n)
// Space: O(1)
class Solution {
    public int jump(int[] nums) {
        int res = 0, l = 0, r = 0, far;

        while(r < nums.length - 1){
            far = 0;
            for(int i = l; i <= r; i++){
                far = Math.max(far, i + nums[i]);
            }
            l = r + 1;
            r = far;
            res += 1;
        }

        return res;
    }
}
